python/mpt/src/mpt/__init__.py

```diff
@@ -0,0 +1,1553 @@
+    code_hash: Bytes32
+
+    def to_account(self, code: Optional[bytes] = None) -> Account:
+        return Account(
+            nonce=Uint(int.from_bytes(self.nonce, "big")),
+            balance=U256(int.from_bytes(self.balance, "big")),
+            code=code,
+        )
+
+    def rlp_encode(self) -> bytes:
+        return rlp.encode([self.nonce, self.balance, self.storage_root, self.code_hash])
+
+
+EMPTY_TRIE_ROOT_HASH = Hash32(
+    bytes.fromhex("56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421")
+)
+
+EMPTY_CODE_HASH = Hash32(
+    bytes.fromhex("c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470")
+)
+
+EMPTY_BYTES_RLP = b"\x80"
+
+
+def nibble_path_to_hex(nibble_path: Bytes) -> str:
+    if len(nibble_path) % 2 != 0:
+        nibble_path = nibble_path + b"\x00"  # Pad with zero if odd
+    result = bytes(
+        [
+            nibble_path[i] * 16 + nibble_path[i + 1]
+            for i in range(0, len(nibble_path), 2)
+        ]
+    )
+    return "0x" + result.hex()
+
+
+class StateTries:
+    """
+    Represents the (partial) state of Ethereum for a given block.
+    Includes all account MPT leaf nodes and storage MPT leaf nodes and all intermediary nodes touched in a block,
+    related codes, an access list of addresses and storage keys that were accessed in the block, and a state root.
+    """
+
+    nodes: Mapping[Bytes32, Bytes]
+    codes: Mapping[Bytes32, Bytes]
+    access_list: Mapping[Address, Optional[List[Bytes32]]]
+    state_root: Hash32
+
+    @classmethod
+    def create_empty(cls) -> "StateTries":
+        return cls(
+            nodes={},
+            codes={},
+            access_list={},
+            state_root=EMPTY_TRIE_ROOT_HASH,
+        )
+
+    def __init__(
+        self,
+        nodes: Mapping[Bytes32, Bytes],
+        codes: Mapping[Bytes32, Bytes],
+        access_list: Mapping[Address, Optional[List[Bytes32]]],
+        state_root: Hash32,
+    ):
+        self.nodes = nodes
+        self.codes = codes
+        self.access_list = access_list
+        self.state_root = state_root
+        logger.debug(f"Initialized MPT with state root: 0x{state_root.hex()}")
+        logger.debug(f"Number of nodes: {len(nodes)}")
+        logger.debug(f"Number of codes: {len(codes)}")
+        logger.debug(f"Number of addresses in access list: {len(access_list)}")
+
+    def to_state(self) -> State:
+        """
+        Convert the StateTries to a State object.
+
+        Returns
+        -------
+        State
+            A State object representing the Ethereum state
+        """
+
+        # Step 1: Recursively explore the state trie to get to the leaves
+        _main_trie: Trie[Address, Optional[Account]] = Trie(
+            secured=True, default=None, _data=defaultdict(lambda: None)
+        )
+        _storage_tries: Dict[Address, Trie[Bytes32, U256]] = dict()
+
+        logger.debug("Starting to convert StateTries to State")
+        for address in self.access_list.keys():
```

Comment:
why do we loop from the access_list and not from the root node?

Reply:
it's lighter to iterate over the access_list 
we could also traverse the trie